src/audio/sound_dose/sound_dose.c

```diff
@@ -0,0 +1,325 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2025 Intel Corporation.
+
+#include <sof/audio/module_adapter/module/generic.h>
+#include <sof/audio/sink_source_utils.h>
+#include <sof/audio/sink_api.h>
+#include <sof/audio/source_api.h>
+#include <sof/math/iir_df1.h>
+#include <kernel/abi.h>
+#include <kernel/header.h>
+#include <rtos/init.h>
+#include <user/sound_dose.h>
+#include <user/eq.h>
+
+#include "sound_dose.h"
+#include "sound_dose_iir_48k.h"
+
+/* UUID identifies the components. Use e.g. command uuidgen from package
+ * uuid-runtime, add it to uuid-registry.txt in SOF top level.
+ */
+SOF_DEFINE_REG_UUID(sound_dose);
+
+/* Creates logging data for the component */
+LOG_MODULE_REGISTER(sound_dose, CONFIG_SOF_LOG_LEVEL);
+
+/* Creates the compont trace. Traces show in trace console the component
+ * info, warning, and error messages.
+ */
+DECLARE_TR_CTX(sound_dose_tr, SOF_UUID(sound_dose_uuid), LOG_LEVEL_INFO);
+
+void sound_dose_report_mel(const struct processing_module *mod)
+{
+	struct sound_dose_comp_data *cd = module_get_private_data(mod);
+	struct comp_dev *dev = mod->dev;
+
+	cd->dose->current_sens_dbfs_dbspl = cd->setup.sens_dbfs_dbspl;
+	cd->dose->current_volume_offset = cd->vol.volume_offset;
+	cd->dose->current_gain = cd->gain;
+	cd->dose->dbfs_value = (int32_t)(((int64_t)cd->level_dbfs * 100) >> 16);
+	cd->dose->mel_value = cd->dose->dbfs_value +
+		cd->setup.sens_dbfs_dbspl +
+		cd->vol.volume_offset;
+	cd->feature->stream_time_ms = (uint64_t)cd->total_frames_count; // TODO convert to time
+
+	comp_info(dev, "MEL %d %d", cd->level_dbfs, ((cd->level_dbfs >> 15) + 1) >> 1);
+}
+
+int sound_dose_filters_init(struct processing_module *mod)
+{
+	struct sound_dose_comp_data *cd = module_get_private_data(mod);
+	struct comp_dev *dev = mod->dev;
+	struct sof_abi_hdr *blob;
+	struct sof_eq_iir_config *iir_config;
+	struct sof_eq_iir_header *iir_coef;
+	size_t iir_size, alloc_size;
+	int32_t *data;
+	int i;
+
+	/* Initialize FIR */
+	switch (cd->rate) {
+	case 48000:
```

Comment:
IIUC SoundDose would be at the end of the playback render pipeline e.g. after any EQ or DRC ? This would mean we should cater for HW render rates like 48k, 96k and 192kHz ?

Reply:
Yep, I should add decimation to loudness measurement for high rates. The A-weight is not defined for high frequencies > 20 kHz, so I assume they can be omitted. The SRC filter and coefficients can be used for the purpose. The MCPS load would also become very high with 192 kHz if not decimated, since the A-weight filter is quite high order.